src/jailer/src/env.rs

```diff
@@ -197,37 +200,53 @@ impl Env {
-        // Unshare into a new PID namespace.
-        // The current process will not be moved into the newly created namespace, but its first
-        // child will assume the role of init(1) in the new namespace.
-        // The call is safe because we're invoking a C library function with valid parameters.
-        SyscallReturnCode(unsafe { libc::unshare(libc::CLONE_NEWPID) })
-            .into_empty_result()
-            .map_err(Error::UnshareNewPID)?;
-
         // Compute jailer's total CPU time up to the current time.
         self.jailer_cpu_time_us =
             utils::time::get_time_us(utils::time::ClockType::ProcessCpu) - self.start_time_cpu_us;
 
-        // Duplicate the current process. The child process will belong to the previously created
-        // PID namespace.
-        // TODO: replace the `unshare()` + `fork()` combo with `clone()` if we ever need to
-        //  squeeze every bit of start-up latency we can get
-        let pid = unsafe { libc::fork() };
-        match pid {
-            0 => {
-                // Reset process start time.
-                self.start_time_cpu_us = 0;
-
-                Err(Error::Exec(self.exec_command(chroot_exec_file)))
-            }
-            child_pid => {
-                // Save the PID of the process running the exec file provided
-                // inside <chroot_exec_file>.pid file.
-                self.save_exec_file_pid(child_pid, chroot_exec_file)?;
-                unsafe { libc::exit(0) }
-            }
+        // This is a wrapper for our child callback to the clone libc function.
+        // We will use clone's arg parameter as the actual callback and call it from here.
+        extern "C" fn on_clone(arg: *mut libc::c_void) -> libc::c_int {
+            let callback_ref = arg as *mut *mut dyn std::ops::FnOnce();
+            let callback = unsafe { Box::from_raw(*callback_ref) };
+
+            callback();
+            0
         }
+
+        let do_on_clone = || {
+            // Reset process start time.
+            self.start_time_cpu_us = 0;
+
+            self.exec_command(chroot_exec_file.clone());
+        };
+
+        let do_on_clone_box: Box<dyn FnOnce()> = Box::new(do_on_clone);
+        let do_on_clone_ref = &mut Box::into_raw(do_on_clone_box);
+
+        // Mapping the child process it's own stack.
+        let mut stack = [0u8; PROCESS_STACK_SIZE];
+
+        // The call is safe because the stack size is a constant positive
+        let stack_top_ptr = stack.last_mut().unwrap() as *mut u8;
+
+        // We will use clone in order to spawn a new child process into a new PID namespace.
+        // The current process will not be moved into the newly created namespace, but the child
+        // spawned via clone will assume the role of init(1) in the new namespace.
+        // The call is safe because we're invoking a C library function with valid parameters.
+        let child_pid = unsafe {
```

Comment:
Should we handle the case where `child_pid` is < 0 ? This would mean that an error occurred.

Reply:
So I did not want to change the original behavior - originally the fork return value was not checked for error either. 
I thought it was done intentionally because then we only save the pid (-1 in an error case) and exit. 
I can change it to be different from the original behavior if you wish so.

Reply:
@luminitavoicu what do you think ? What was the reasoning behind this when implementing `fork()` ?

Reply:
Luminita is OOTO a couple of days so we'll probably have her input next week. Anyway personally I think we should return an Err when `child_pid` is < 0 . It would make it easier to spot a problem.

Reply:
I am in favor of returning an error here if the child PID is negative, as it would make it easier to debug possible problems.